### PR TITLE
SCUMM: SMUSH (FT/DIG demo): Rework audio sub-engine 

### PR DESCRIPTION
**Pull request by AndywinXp**
During some other research for other stuff I noticed that in SMUSH we're currently playing audio without handling the flags. Turns out that was only the tip of the iceberg: there's a whole miniature audio engine inside SMUSH (for FT and DIG demo) which handles PSAD chunk quite differently than we're doing right now.

This PR replaces the old PSAD sound frame handler with the one used in the FT interpreter, and in consequence fixes [#13322](https://bugs.scummvm.org/ticket/13322). Also, this adds a couple of features we didn't have before: chunks crossfading and music volume reduction during speech.

I'm currently testing FT with this new engine and everything seems to work fine; I want to take some time in the next days to compare the code against FT demo and DIG demo disasms, to see if I missed something else, but the bulk of it should be complete and stable.

~~As usual, even though this is WIP (not for long I hope) but still fully functional, I wanted to publish this PR to get some opinions on it.~~

EDIT: Hey, PR number 3800, round number!

### CODE REVIEW

File: engines/scumm/smush/smush_player.cpp [hunk 1]
```diff
@@ -1304,4 +1320,856 @@ void SmushPlayer::play(const char *filename, int32 speed, int32 offset, int32 st
 	CursorMan.showMouse(oldMouseState);
 }
 
+void SmushPlayer::initAudio(int samplerate, int32 maxChunkSize, int32 track1MaxSize, int32 track2MaxSize, int32 track3MaxSize, int32 track4MaxSize) {
+	char *v5;
```
Comment (aquadran) [inline engines/scumm/smush/smush_player.cpp]: don't put not ready code to PR

**Comment by AndywinXp** [comment]
Hmm, it appears I have broken something :-) too tired to look at it now, I will try and fix it tomorrow

File: engines/scumm/smush/smush_player.cpp (hunk 1 above)
Comment (AndywinXp) [reply engines/scumm/smush/smush_player.cpp]: Okay, it appears I haven't commited part of the PR. Sorry for that. I'll do that and rebase pretty quick.

Comment (aquadran) [reply engines/scumm/smush/smush_player.cpp]: I think it's better do some squash ,commits contains some commented draft code

Comment (AndywinXp) [reply engines/scumm/smush/smush_player.cpp]: You are certainly right; done

File: engines/scumm/smush/smush_player.cpp [hunk 2]
```diff
@@ -1304,4 +1279,770 @@ void SmushPlayer::play(const char *filename, int32 speed, int32 offset, int32 st
...
+		// Find the maximum block sizes between the current ones
+		maxBlockSize = 0;
+		for (int i = 0; i < _smushNumTracks; i++) {
+			if (_smushTracks[i].blockSize > maxBlockSize) {
+				maxBlockSize = _smushTracks[i].blockSize;
+			}
+		}
+
+		diff2 = 10000000;
+		if ((flags & 0xC0) != IS_BKG_MUSIC) {
```
Comment (aquadran) [inline engines/scumm/smush/smush_player.cpp:1415]: if you using macro, so I guess both should have defined macro

File: engines/scumm/smush/smush_player.cpp [hunk 3]
```diff
@@ -1304,4 +1279,770 @@ void SmushPlayer::play(const char *filename, int32 speed, int32 offset, int32 st
...
+					fadeInFrameCount = (fadeRemaining < DIMUSE_FEEDSIZE) ? fadeRemaining : DIMUSE_FEEDSIZE;
+
+					if (fadeInFrameCount == maxFadeChunkSize) {
+						fadeFeedSize = feedSize;
+					} else {
+						fadeFeedSize = _smushAudioSampleRate * fadeInFrameCount / _smushDispatch[i].fadeSampleRate;
+					}
+
+					if (isPlayableTrack) {
+						fadeVolume = _smushDispatch[i].fadeRemaining * _smushDispatch[i].fadeVolume * _smushTrackVols[0] / 0x5F400;
```
Comment (aquadran) [inline engines/scumm/smush/smush_player.cpp:1661]: dividing by hex looks weird, as it's not round number :)

File: engines/scumm/smush/smush_player.cpp (hunk 2 above)
Comment (AndywinXp) [reply engines/scumm/smush/smush_player.cpp:1415]: Okay, so you mean something like `(flags & TRK_TYPE_FLAG)`?

File: engines/scumm/smush/smush_player.cpp [hunk 4]
```diff
@@ -1304,4 +1279,770 @@ void SmushPlayer::play(const char *filename, int32 speed, int32 offset, int32 st
...
+}
+
+void SmushPlayer::feedAudio(uint8 *srcBuf, int groupId, int volume, int pan, int16 flags) {
+	int panDelta, effPan;
+	int32 maxFrames;
+	uint16 trkId, index;
+
+	if (_smushAudioInitialized) {
+		// Check file encoding
+		if (!srcBuf[8] && !srcBuf[9] && !srcBuf[12] && !srcBuf[13] && !srcBuf[16] && !srcBuf[17]) {
```
Comment (aquadran) [inline engines/scumm/smush/smush_player.cpp:2001]: preferable I would see "srcBuf[8] != 0" bytes are no boolean type, casting to Boolean.  this way is rather C thing, not C++

Comment (AndywinXp) [reply engines/scumm/smush/smush_player.cpp:2001]: Thanks, I didn't know that, I will correct that accordingly

File: engines/scumm/smush/smush_player.cpp (hunk 3 above)
Comment (AndywinXp) [reply engines/scumm/smush/smush_player.cpp:1661]: Hmm, you're right. I'll see if I can find a convincing short name for turning that into a macro, like SMUSH_MAX_FADE_VOL, and convert it to base 10 anyway.

File: engines/scumm/smush/smush_player.cpp (hunk 2 above)
Comment (AndywinXp) [reply engines/scumm/smush/smush_player.cpp:1415]: Okay, I will do that, thanks

**Comment by AndywinXp** [comment]
The code is now feature complete and has been checked against the other disasms (even though I still have to finish testing the games). Should it be moved to a separate `smush_audio.cpp` file along with the audio handling for other SCUMM v7-8 games?

**Comment by aquadran** [comment]
> The code is now feature complete and has been checked against the other disasms (even though I still have to finish testing the games). Should it be moved to a separate `smush_audio.cpp` file along with the audio handling for other SCUMM v7-8 games?

depends how much is different. if there are few differences keep in one file. too much better separated file

**Comment by AndywinXp** [comment]
> depends how much is different. if there are few differences keep in one file. too much better separated file
Okay; this PR replaces the whole audio subsystem for FT and is a little bit smaller than the previous iteration since it relies on the iMUSE mixer instead of its own personal channel/mixer abstraction. I personally think it would be better to have a separate file, and (given this info) if you think the same I'll do the code separation.

File: engines/scumm/smush/smush_player.cpp [hunk 5]
```diff
@@ -1304,4 +1278,762 @@ void SmushPlayer::play(const char *filename, int32 speed, int32 offset, int32 st
...
+					_smushTracks[i].state == TRK_STATE_ENDING   ||
+					_smushTracks[i].flags <= flags) {
+					if (_smushTracks[i].state == TRK_STATE_INACTIVE)
+						flagsAccumulator[i] = 0x1000;
+
+					if (_smushTracks[i].state == TRK_STATE_ENDING)
+						flagsAccumulator[i] += 0x200;
+
+					if (_smushTracks[i].flags == flags)
+						flagsAccumulator[i] += 0x400;
```
Comment (aquadran) [inline engines/scumm/smush/smush_player.cpp:1441]: why flags use "+" ?

Comment (AndywinXp) [reply engines/scumm/smush/smush_player.cpp:1441]: It's a mislabeling on my end, I guess I have to find a better name for that array which is an accumulator whose value depends on the flags set  :)

**Comment by AndywinXp** [comment]
All games affected by these changes are tested and the audio works as expected. As a plus, given that the compressed audio flow is separate from this one, it works fine just as it did before. I'm now ready for a final code review.

**Review by aquadran — APPROVED** [review]

**Comment by AndywinXp** [comment]
Thanks for the review. Merging!

**Pull request merged**